File: csharp/LicenseCM.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.NetworkInformation;
using System.Runtime.InteropServices;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using System.Timers;

namespace LicenseCM
{
    public class SecurityViolationEventArgs : EventArgs
    {
        public string Type { get; set; }
        public string Reason { get; set; }
        public object Details { get; set; }
    }

    public class HeartbeatFailedEventArgs : EventArgs
    {
        public string Error { get; set; }
    }

    public class LicenseCMClient : IDisposable
    {
        private readonly string _baseUrl;
        private readonly string _productId;
        private readonly string _secretKey;
        private readonly bool _useEncryption;
        private readonly bool _autoHeartbeat;
        private readonly int _heartbeatIntervalMs;
        private readonly HttpClient _httpClient;

        // Session state
        private string _sessionToken;
        private DateTime? _sessionExpires;
        private System.Timers.Timer _heartbeatTimer;
        private string _licenseKey;
        private string _hwid;

        // Public key for signature verification
        private string _publicKey;

        // Events
        public event EventHandler SessionExpired;
        public event EventHandler<SecurityViolationEventArgs> SecurityViolation;
        public event EventHandler<HeartbeatFailedEventArgs> HeartbeatFailed;

        public LicenseCMClient(
            string baseUrl,
            string productId,
            string secretKey = null,
            bool useEncryption = false,
            bool autoHeartbeat = true,
            int heartbeatIntervalMs = 300000) // 5 minutes
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _productId = productId;
            _secretKey = secretKey;
            _useEncryption = useEncryption && !string.IsNullOrEmpty(secretKey);
            _autoHeartbeat = autoHeartbeat;
            _heartbeatIntervalMs = heartbeatIntervalMs;
            _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
        }

        #region HWID Generation

        public static string GenerateHwid()
        {
            var components = new List<string>
            {
                Environment.OSVersion.Platform.ToString(),
                RuntimeInformation.OSArchitecture.ToString(),
                Environment.MachineName,
                Environment.ProcessorCount.ToString()
            };

            // Get MAC address
            try
            {
                var mac = NetworkInterface.GetAllNetworkInterfaces()
                    .Where(nic => nic.OperationalStatus == OperationalStatus.Up &&
                                  nic.NetworkInterfaceType != NetworkInterfaceType.Loopback)
                    .Select(nic => nic.GetPhysicalAddress().ToString())
                    .FirstOrDefault();

                if (!string.IsNullOrEmpty(mac))
                    components.Add(mac);
            }
            catch { }

            // Get disk serial (Windows only)
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                try
                {
                    var psi = new ProcessStartInfo
                    {
                        FileName = "wmic",
                        Arguments = "diskdrive get serialnumber",
                        RedirectStandardOutput = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    };

                    using var process = Process.Start(psi);
                    var output = process.StandardOutput.ReadToEnd();
                    var lines = output.Split('\n');
                    if (lines.Length > 1)
                    {
                        var serial = lines[1].Trim();
                        if (!string.IsNullOrEmpty(serial))
                            components.Add(serial);
                    }
                }
                catch { }
            }

            var data = string.Join("|", components);
            using var sha256 = SHA256.Create();
            var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(data));
            return BitConverter.ToString(hash).Replace("-", "").ToLower();
        }

        #endregion

        #region Security Detection

        private Dictionary<string, object> CollectClientData()
        {
            var data = new Dictionary<string, object>
            {
                ["hwid"] = _hwid ?? GenerateHwid(),
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                ["platform"] = Environment.OSVersion.Platform.ToString(),
                ["os_version"] = Environment.OSVersion.VersionString,
                ["architecture"] = RuntimeInformation.OSArchitecture.ToString(),
                ["hostname"] = Environment.MachineName,
                ["dotnet_version"] = RuntimeInformation.FrameworkDescription,
                ["cpu_count"] = Environment.ProcessorCount,
                ["env_indicators"] = new Dictionary<string, bool>
                {
                    ["debug_mode"] = Debugger.IsAttached,
                    ["is_64bit_process"] = Environment.Is64BitProcess,
                    ["is_64bit_os"] = Environment.Is64BitOperatingSystem
                },
                ["vm_indicators"] = DetectVMIndicators(),
                ["debug_indicators"] = DetectDebugIndicators()
            };

            return data;
        }

        private List<string> DetectVMIndicators()
        {
            var indicators = new List<string>();
            var hostname = Environment.MachineName.ToLower();

            // Check hostname patterns
            var vmHostnames = new[] { "vmware", "virtualbox", "sandbox", "virtual", "qemu" };
            if (vmHostnames.Any(vm => hostname.Contains(vm)))
                indicators.Add("suspicious_hostname");

            // Check MAC address prefixes
            var vmMacPrefixes = new[]
            {
                "000C29", "005056", "000569", // VMware
                "080027", "0A0027",           // VirtualBox
                "00155D",                      // Hyper-V
                "00163E",                      // Xen
                "525400"                       // QEMU
            };

            try
            {
                var macs = NetworkInterface.GetAllNetworkInterfaces()
                    .Where(nic => nic.NetworkInterfaceType != NetworkInterfaceType.Loopback)
                    .Select(nic => nic.GetPhysicalAddress().ToString().ToUpper());

                if (macs.Any(mac => vmMacPrefixes.Any(prefix => mac.StartsWith(prefix))))
                    indicators.Add("vm_mac_address");
            }
            catch { }

            // Check for low CPU count
            if (Environment.ProcessorCount < 2)
                indicators.Add("single_cpu");

            return indicators;
        }

        private List<string> DetectDebugIndicators()
        {
            var indicators = new List<string>();

            // Check if debugger is attached
            if (Debugger.IsAttached)
                indicators.Add("debugger_attached");

            // Timing analysis
            var sw = Stopwatch.StartNew();
            for (int i = 0; i < 1000; i++)
            {
                _ = Guid.NewGuid();
            }
            sw.Stop();

            if (sw.ElapsedMilliseconds > 100)
                indicators.Add("timing_anomaly");

            // Check for debug environment variables
            if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable("DEBUG")))
                indicators.Add("env_debug");

            return indicators;
        }

        #endregion

        #region Encryption

        private (string iv, string data, string tag) Encrypt(string plaintext)
        {
            using var aes = new AesGcm(Encoding.UTF8.GetBytes(_secretKey.PadRight(32).Substring(0, 32)));
            var iv = new byte[16];
            RandomNumberGenerator.Fill(iv);

            var plaintextBytes = Encoding.UTF8.GetBytes(plaintext);
            var ciphertext = new byte[plaintextBytes.Length];
            var tag = new byte[16];

            aes.Encrypt(iv, plaintextBytes, ciphertext, tag);

            return (
                BitConverter.ToString(iv).Replace("-", "").ToLower(),
                BitConverter.ToString(ciphertext).Replace("-", "").ToLower(),
                BitConverter.ToString(tag).Replace("-", "").ToLower()
            );
        }

        private string Decrypt(string iv, string data, string tag)
        {
            using var aes = new AesGcm(Encoding.UTF8.GetBytes(_secretKey.PadRight(32).Substring(0, 32)));
            var ivBytes = HexToBytes(iv);
            var ciphertext = HexToBytes(data);
            var tagBytes = HexToBytes(tag);
            var plaintext = new byte[ciphertext.Length];

            aes.Decrypt(ivBytes, ciphertext, tagBytes, plaintext);

            return Encoding.UTF8.GetString(plaintext);
        }

        private string Sign(string data)
        {
            using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(_secretKey));
            var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
            return BitConverter.ToString(hash).Replace("-", "").ToLower();
        }

        private byte[] HexToBytes(string hex)
        {
            var bytes = new byte[hex.Length / 2];
            for (int i = 0; i < bytes.Length; i++)
            {
                bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
            }
            return bytes;
        }

        #endregion

        #region Public Key

        public async Task<string> FetchPublicKeyAsync()
        {
            try
            {
                var response = await _httpClient.GetAsync($"{_baseUrl}/api/client/public-key");
                var content = await response.Content.ReadAsStringAsync();
                var json = JsonSerializer.Deserialize<JsonElement>(content);

                if (json.GetProperty("success").GetBoolean())
                {
                    _publicKey = json.GetProperty("data").GetProperty("public_key").GetString();
                    return _publicKey;
                }
            }
            catch { }

            return null;
        }

        #endregion

        #region API Requests

        private async Task<JsonElement> RequestAsync(string endpoint, string licenseKey, string hwid)
        {
            var clientData = CollectClientData();

            var body = new Dictionary<string, object>
            {
                ["license_key"] = licenseKey,
                ["hwid"] = hwid,
                ["product_id"] = _productId,
                ["client_data"] = clientData
            };

            if (!string.IsNullOrEmpty(_sessionToken))
                body["session_token"] = _sessionToken;

            string jsonBody;

            if (_useEncryption)
            {
                var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
                var encrypted = Encrypt(JsonSerializer.Serialize(body));
                var signaturePayload = $"{encrypted.iv}:{encrypted.data}:{encrypted.tag}:{timestamp}";
                var signature = Sign(signaturePayload);

                var encryptedBody = new Dictionary<string, object>
                {
                    ["encrypted"] = true,
                    ["iv"] = encrypted.iv,
                    ["data"] = encrypted.data,
                    ["tag"] = encrypted.tag,
                    ["signature"] = signature,
                    ["product_id"] = _productId,
                    ["timestamp"] = timestamp
                };
                jsonBody = JsonSerializer.Serialize(encryptedBody);
            }
            else
            {
                jsonBody = JsonSerializer.Serialize(body);
            }

            var content = new StringContent(jsonBody, Encoding.UTF8, "application/json");
            var response = await _httpClient.PostAsync($"{_baseUrl}/api/client{endpoint}", content);
            var responseBody = await response.Content.ReadAsStringAsync();

            var responseJson = JsonSerializer.Deserialize<JsonElement>(responseBody);

            if (_useEncryption && responseJson.TryGetProperty("encrypted", out var enc) && enc.GetBoolean())
            {
                var decrypted = Decrypt(
                    responseJson.GetProperty("iv").GetString(),
                    responseJson.GetProperty("data").GetString(),
                    responseJson.GetProperty("tag").GetString()
                );
                responseJson = JsonSerializer.Deserialize<JsonElement>(decrypted);
            }

            if (responseJson.GetProperty("success").GetBoolean())
            {
                var data = responseJson.GetProperty("data");

                // Handle session token rotation
                if (data.TryGetProperty("new_token", out var newToken))
                    _sessionToken = newToken.GetString();

                // Handle session info
                if (data.TryGetProperty("session", out var session))
                {
                    _sessionToken = session.GetProperty("token").GetString();
                    if (session.TryGetProperty("expires_at", out var expiresAt))
                        _sessionExpires = DateTime.Parse(expiresAt.GetString());
                }

                return data;
            }
            else
            {
                // Handle security violations
                if (responseJson.TryGetProperty("security_blocked", out var blocked) && blocked.GetBoolean())
                {
                    SecurityViolation?.Invoke(this, new SecurityViolationEventArgs
                    {
                        Type = "blocked",
                        Reason = responseJson.GetProperty("message").GetString()
                    });
                }

                throw new Exception(responseJson.GetProperty("message").GetString());
            }
        }

        #endregion

        #region License Operations

        public async Task InitializeAsync()
        {
            await FetchPublicKeyAsync();
        }

        public async Task<JsonElement> ValidateAsync(string licenseKey, string hwid = null)
        {
            _licenseKey = licenseKey;
            _hwid = hwid ?? GenerateHwid();
            return await RequestAsync("/validate", licenseKey, _hwid);
        }

        public async Task<JsonElement> ActivateAsync(string licenseKey, string hwid = null)
        {
            _licenseKey = licenseKey;
            _hwid = hwid ?? GenerateHwid();

            var result = await RequestAsync("/activate", licenseKey, _hwid);

            // Start heartbeat if enabled
            if (_autoHeartbeat && result.TryGetProperty("session", out _))
                StartHeartbeat();

            return result;
        }

        public async Task<JsonElement> DeactivateAsync(string licenseKey = null, string hwid = null)
        {
            StopHeartbeat();

            var result = await RequestAsync("/deactivate",
                licenseKey ?? _licenseKey,
                hwid ?? _hwid ?? GenerateHwid());

            _sessionToken = null;
            _sessionExpires = null;

            return result;
        }

        public async Task<JsonElement> HeartbeatAsync(string licenseKey = null, string hwid = null)
        {
            return await RequestAsync("/heartbeat",
                licenseKey ?? _licenseKey,
                hwid ?? _hwid ?? GenerateHwid());
        }

        public async Task<JsonElement> VerifyChallengeAsync(string challenge)
        {
            var hwid = _hwid ?? GenerateHwid();
            using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(hwid));
            var response = BitConverter.ToString(
                hmac.ComputeHash(Encoding.UTF8.GetBytes(challenge))
            ).Replace("-", "").ToLower();

            var body = new Dictionary<string, object>
            {
                ["license_key"] = _licenseKey,
                ["hwid"] = hwid,
                ["challenge"] = challenge,
                ["response"] = response,
                ["product_id"] = _productId
            };

            var jsonBody = JsonSerializer.Serialize(body);
            var content = new StringContent(jsonBody, Encoding.UTF8, "application/json");
            var httpResponse = await _httpClient.PostAsync($"{_baseUrl}/api/client/verify-challenge", content);
            var responseBody = await httpResponse.Content.ReadAsStringAsync();

            return JsonSerializer.Deserialize<JsonElement>(responseBody);
        }

        #endregion

        #region Heartbeat Management

        public void StartHeartbeat()
        {
            StopHeartbeat();

            _heartbeatTimer = new System.Timers.Timer(_heartbeatIntervalMs);
            _heartbeatTimer.Elapsed += async (s, e) => await HeartbeatLoopAsync();
            _heartbeatTimer.AutoReset = true;
            _heartbeatTimer.Start();
        }

        public void StopHeartbeat()
        {
            _heartbeatTimer?.Stop();
            _heartbeatTimer?.Dispose();
            _heartbeatTimer = null;
        }

        private async Task HeartbeatLoopAsync()
        {
            try
            {
                await HeartbeatAsync();
            }
            catch (Exception ex)
            {
                HeartbeatFailed?.Invoke(this, new HeartbeatFailedEventArgs { Error = ex.Message });

                if (ex.Message.ToLower().Contains("expired") || ex.Message.ToLower().Contains("invalid"))
                {
                    StopHeartbeat();
                    SessionExpired?.Invoke(this, EventArgs.Empty);
                }
            }
        }

        #endregion

        #region Session Info

        public bool IsSessionValid()
        {
            return !string.IsNullOrEmpty(_sessionToken) &&
                   _sessionExpires.HasValue &&
                   DateTime.UtcNow < _sessionExpires.Value;
        }

        public (string Token, DateTime? Expires, bool IsValid) GetSessionInfo()
        {
            return (_sessionToken, _sessionExpires, IsSessionValid());
        }

        #endregion

        #region Disposal

        public void Dispose()
        {
            StopHeartbeat();
            _httpClient?.Dispose();
            _sessionToken = null;
            _sessionExpires = null;
            _licenseKey = null;
            _hwid = null;
        }

        #endregion
    }

    // Example usage
    class Program
    {
        static async Task Main(string[] args)
        {
            using var client = new LicenseCMClient(
                baseUrl: "http://localhost:3000",
                productId: "your-product-id",
                secretKey: "your-secret-key",
                useEncryption: true,
                autoHeartbeat: true,
                heartbeatIntervalMs: 300000 // 5 minutes
            );

            // Event handlers
            client.SessionExpired += (s, e) =>
            {
                Console.WriteLine("Session expired! Please re-activate.");
                Environment.Exit(1);
            };

            client.SecurityViolation += (s, e) =>
            {
                Console.WriteLine($"Security violation: {e.Type} - {e.Reason}");
                Environment.Exit(1);
            };

            client.HeartbeatFailed += (s, e) =>
            {
                Console.WriteLine($"Heartbeat failed: {e.Error}");
            };

            string licenseKey = "XXXX-XXXX-XXXX-XXXX";

            try
            {
                // Initialize (fetch public key)
                await client.InitializeAsync();

                // Activate license
                var result = await client.ActivateAsync(licenseKey);
                Console.WriteLine($"License activated: {result}");

                // License is now active with automatic heartbeat
                // The client will send heartbeats every 5 minutes

                // Keep running
                Console.WriteLine("Press Enter to exit...");
                Console.ReadLine();

                // Cleanup
                await client.DeactivateAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Activation failed: {ex.Message}");
            }
        }
    }
}

File: go/licensecm.go
// LicenseCM Go SDK with Enhanced Security Features
package licensecm

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"sync"
	"time"
)

// Client represents the LicenseCM client
type Client struct {
	BaseURL           string
	ProductID         string
	SecretKey         string
	UseEncryption     bool
	AutoHeartbeat     bool
	HeartbeatInterval time.Duration

	// Session state
	sessionToken   string
	sessionExpires *time.Time
	heartbeatStop  chan struct{}
	licenseKey     string
	hwid           string
	publicKey      string
	mu             sync.Mutex

	// Callbacks
	OnSessionExpired    func()
	OnSecurityViolation func(details map[string]interface{})
	OnHeartbeatFailed   func(err error)

	httpClient *http.Client
}

// NewClient creates a new LicenseCM client
func NewClient(baseURL, productID, secretKey string) *Client {
	return &Client{
		BaseURL:           strings.TrimRight(baseURL, "/"),
		ProductID:         productID,
		SecretKey:         secretKey,
		UseEncryption:     false,
		AutoHeartbeat:     true,
		HeartbeatInterval: 5 * time.Minute,
		httpClient:        &http.Client{Timeout: 30 * time.Second},
		OnSessionExpired:  func() {},
		OnSecurityViolation: func(details map[string]interface{}) {},
		OnHeartbeatFailed: func(err error) {},
	}
}

// GenerateHWID generates a hardware ID from system info
func GenerateHWID() string {
	var components []string

	// Platform info
	components = append(components, runtime.GOOS, runtime.GOARCH)

	// Hostname
	if hostname, err := os.Hostname(); err == nil {
		components = append(components, hostname)
	}

	// MAC address
	if interfaces, err := net.Interfaces(); err == nil {
		for _, iface := range interfaces {
			if iface.Flags&net.FlagLoopback == 0 && len(iface.HardwareAddr) > 0 {
				components = append(components, iface.HardwareAddr.String())
				break
			}
		}
	}

	// CPU count
	components = append(components, fmt.Sprintf("%d", runtime.NumCPU()))

	// Disk serial (Windows)
	if runtime.GOOS == "windows" {
		if out, err := exec.Command("wmic", "diskdrive", "get", "serialnumber").Output(); err == nil {
			lines := strings.Split(string(out), "\n")
			if len(lines) > 1 {
				serial := strings.TrimSpace(lines[1])
				if serial != "" {
					components = append(components, serial)
				}
			}
		}
	}

	data := strings.Join(components, "|")
	hash := sha256.Sum256([]byte(data))
	return hex.EncodeToString(hash[:])
}

// ClientData represents client environment data
type ClientData struct {
	HWID            string                 `json:"hwid"`
	Timestamp       int64                  `json:"timestamp"`
	Platform        string                 `json:"platform"`
	Architecture    string                 `json:"architecture"`
	Hostname        string                 `json:"hostname"`
	GoVersion       string                 `json:"go_version"`
	CPUCount        int                    `json:"cpu_count"`
	EnvIndicators   map[string]interface{} `json:"env_indicators"`
	VMIndicators    []string               `json:"vm_indicators"`
	DebugIndicators []string               `json:"debug_indicators"`
}

func (c *Client) collectClientData() ClientData {
	hostname, _ := os.Hostname()

	data := ClientData{
		HWID:         c.hwid,
		Timestamp:    time.Now().UnixMilli(),
		Platform:     runtime.GOOS,
		Architecture: runtime.GOARCH,
		Hostname:     hostname,
		GoVersion:    runtime.Version(),
		CPUCount:     runtime.NumCPU(),
		EnvIndicators: map[string]interface{}{
			"debug_mode":  os.Getenv("DEBUG") != "",
			"go_debug":    os.Getenv("GODEBUG") != "",
			"num_goroutines": runtime.NumGoroutine(),
		},
		VMIndicators:    c.detectVMIndicators(),
		DebugIndicators: c.detectDebugIndicators(),
	}

	if data.HWID == "" {
		data.HWID = GenerateHWID()
	}

	return data
}

func (c *Client) detectVMIndicators() []string {
	var indicators []string

	hostname, _ := os.Hostname()
	hostnameLower := strings.ToLower(hostname)

	// Check hostname patterns
	vmHostnames := []string{"vmware", "virtualbox", "sandbox", "virtual", "qemu"}
	for _, vm := range vmHostnames {
		if strings.Contains(hostnameLower, vm) {
			indicators = append(indicators, "suspicious_hostname")
			break
		}
	}

	// Check MAC address prefixes
	vmMacPrefixes := []string{
		"00:0c:29", "00:50:56", "00:05:69", // VMware
		"08:00:27", "0a:00:27", // VirtualBox
		"00:15:5d", // Hyper-V
		"00:16:3e", // Xen
		"52:54:00", // QEMU
	}

	if interfaces, err := net.Interfaces(); err == nil {
		for _, iface := range interfaces {
			mac := strings.ToLower(iface.HardwareAddr.String())
			for _, prefix := range vmMacPrefixes {
				if strings.HasPrefix(mac, prefix) {
					indicators = append(indicators, "vm_mac_address")
					break
				}
			}
		}
	}

	// Check CPU count
	if runtime.NumCPU() < 2 {
		indicators = append(indicators, "single_cpu")
	}

	return indicators
}

func (c *Client) detectDebugIndicators() []string {
	var indicators []string

	// Check environment variables
	debugEnvVars := []string{"DEBUG", "GODEBUG"}
	for _, env := range debugEnvVars {
		if os.Getenv(env) != "" {
			indicators = append(indicators, "env_"+strings.ToLower(env))
		}
	}

	// Timing analysis
	start := time.Now()
	for i := 0; i < 1000; i++ {
		_ = time.Now().UnixNano()
	}
	duration := time.Since(start)

	if duration.Milliseconds() > 100 {
		indicators = append(indicators, "timing_anomaly")
	}

	return indicators
}

// Encrypt encrypts data using AES-256-GCM
func (c *Client) encrypt(data interface{}) (map[string]string, error) {
	key := []byte(c.SecretKey)
	if len(key) > 32 {
		key = key[:32]
	} else if len(key) < 32 {
		key = append(key, make([]byte, 32-len(key))...)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	iv := make([]byte, 16)
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, err
	}

	plaintext, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	ciphertext := gcm.Seal(nil, iv[:gcm.NonceSize()], plaintext, nil)

	// Separate auth tag
	tagStart := len(ciphertext) - gcm.Overhead()
	tag := ciphertext[tagStart:]
	ciphertext = ciphertext[:tagStart]

	return map[string]string{
		"iv":   hex.EncodeToString(iv),
		"data": hex.EncodeToString(ciphertext),
		"tag":  hex.EncodeToString(tag),
	}, nil
}

// Decrypt decrypts data using AES-256-GCM
func (c *Client) decrypt(encrypted map[string]interface{}) (map[string]interface{}, error) {
	key := []byte(c.SecretKey)
	if len(key) > 32 {
		key = key[:32]
	} else if len(key) < 32 {
		key = append(key, make([]byte, 32-len(key))...)
	}

	iv, _ := hex.DecodeString(encrypted["iv"].(string))
	ciphertext, _ := hex.DecodeString(encrypted["data"].(string))
	tag, _ := hex.DecodeString(encrypted["tag"].(string))

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	// Append tag to ciphertext
	ciphertext = append(ciphertext, tag...)

	plaintext, err := gcm.Open(nil, iv[:gcm.NonceSize()], ciphertext, nil)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	err = json.Unmarshal(plaintext, &result)
	return result, err
}

// Sign creates HMAC-SHA256 signature
func (c *Client) sign(data string) string {
	h := hmac.New(sha256.New, []byte(c.SecretKey))
	h.Write([]byte(data))
	return hex.EncodeToString(h.Sum(nil))
}

// FetchPublicKey fetches the public key from server
func (c *Client) FetchPublicKey() error {
	resp, err := c.httpClient.Get(c.BaseURL + "/api/client/public-key")
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return err
	}

	if success, ok := result["success"].(bool); ok && success {
		if data, ok := result["data"].(map[string]interface{}); ok {
			if pk, ok := data["public_key"].(string); ok {
				c.publicKey = pk
				return nil
			}
		}
	}

	return errors.New("failed to fetch public key")
}

// Initialize initializes the client
func (c *Client) Initialize() error {
	return c.FetchPublicKey()
}

// Request makes an API request
func (c *Client) request(endpoint string, data map[string]interface{}) (map[string]interface{}, error) {
	clientData := c.collectClientData()

	body := map[string]interface{}{
		"product_id":  c.ProductID,
		"client_data": clientData,
	}

	for k, v := range data {
		body[k] = v
	}

	c.mu.Lock()
	if c.sessionToken != "" {
		body["session_token"] = c.sessionToken
	}
	c.mu.Unlock()

	var requestBody []byte
	var err error

	if c.UseEncryption && c.SecretKey != "" {
		timestamp := time.Now().UnixMilli()
		encrypted, err := c.encrypt(body)
		if err != nil {
			return nil, err
		}

		signaturePayload := fmt.Sprintf("%s:%s:%s:%d",
			encrypted["iv"], encrypted["data"], encrypted["tag"], timestamp)
		signature := c.sign(signaturePayload)

		encryptedBody := map[string]interface{}{
			"encrypted":  true,
			"iv":         encrypted["iv"],
			"data":       encrypted["data"],
			"tag":        encrypted["tag"],
			"signature":  signature,
			"product_id": c.ProductID,
			"timestamp":  timestamp,
		}
		requestBody, err = json.Marshal(encryptedBody)
	} else {
		requestBody, err = json.Marshal(body)
	}

	if err != nil {
		return nil, err
	}

	resp, err := c.httpClient.Post(
		c.BaseURL+"/api/client"+endpoint,
		"application/json",
		bytes.NewReader(requestBody),
	)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var response map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return nil, err
	}

	// Decrypt if encrypted
	if encrypted, ok := response["encrypted"].(bool); ok && encrypted && c.UseEncryption {
		response, err = c.decrypt(response)
		if err != nil {
			return nil, err
		}
	}

	if success, ok := response["success"].(bool); ok && success {
		data := response["data"].(map[string]interface{})

		c.mu.Lock()
		// Handle session token rotation
		if newToken, ok := data["new_token"].(string); ok {
			c.sessionToken = newToken
		}

		// Handle session info
		if session, ok := data["session"].(map[string]interface{}); ok {
			if token, ok := session["token"].(string); ok {
				c.sessionToken = token
			}
			if expiresAt, ok := session["expires_at"].(string); ok {
				if t, err := time.Parse(time.RFC3339, expiresAt); err == nil {
					c.sessionExpires = &t
				}
			}
		}
		c.mu.Unlock()

		return data, nil
	}

	// Handle security violations
	if blocked, ok := response["security_blocked"].(bool); ok && blocked {
		c.OnSecurityViolation(map[string]interface{}{
			"type":   "blocked",
			"reason": response["message"],
		})
	}

	return nil, errors.New(response["message"].(string))
}

// Validate validates a license
func (c *Client) Validate(licenseKey string, hwid string) (map[string]interface{}, error) {
	c.licenseKey = licenseKey
	if hwid == "" {
		hwid = GenerateHWID()
	}
	c.hwid = hwid

	return c.request("/validate", map[string]interface{}{
		"license_key": licenseKey,
		"hwid":        hwid,
	})
}

// Activate activates a license
func (c *Client) Activate(licenseKey string, hwid string) (map[string]interface{}, error) {
	c.licenseKey = licenseKey
	if hwid == "" {
		hwid = GenerateHWID()
	}
	c.hwid = hwid

	result, err := c.request("/activate", map[string]interface{}{
		"license_key": licenseKey,
		"hwid":        hwid,
	})

	if err == nil && c.AutoHeartbeat {
		if _, ok := result["session"]; ok {
			c.StartHeartbeat()
		}
	}

	return result, err
}

// Deactivate deactivates a license
func (c *Client) Deactivate(licenseKey, hwid string) (map[string]interface{}, error) {
	c.StopHeartbeat()

	if licenseKey == "" {
		licenseKey = c.licenseKey
	}
	if hwid == "" {
		hwid = c.hwid
		if hwid == "" {
			hwid = GenerateHWID()
		}
	}

	result, err := c.request("/deactivate", map[string]interface{}{
		"license_key": licenseKey,
		"hwid":        hwid,
	})

	c.mu.Lock()
	c.sessionToken = ""
	c.sessionExpires = nil
	c.mu.Unlock()

	return result, err
}

// Heartbeat sends a heartbeat
func (c *Client) Heartbeat(licenseKey, hwid string) (map[string]interface{}, error) {
	if licenseKey == "" {
		licenseKey = c.licenseKey
	}
	if hwid == "" {
		hwid = c.hwid
		if hwid == "" {
			hwid = GenerateHWID()
		}
	}

	return c.request("/heartbeat", map[string]interface{}{
		"license_key": licenseKey,
		"hwid":        hwid,
	})
}

// StartHeartbeat starts automatic heartbeat
func (c *Client) StartHeartbeat() {
	c.StopHeartbeat()

	c.heartbeatStop = make(chan struct{})

	go func() {
		ticker := time.NewTicker(c.HeartbeatInterval)
		defer ticker.Stop()

		for {
			select {
			case <-ticker.C:
				if _, err := c.Heartbeat("", ""); err != nil {
					c.OnHeartbeatFailed(err)

					errMsg := strings.ToLower(err.Error())
					if strings.Contains(errMsg, "expired") || strings.Contains(errMsg, "invalid") {
						c.StopHeartbeat()
						c.OnSessionExpired()
						return
					}
				}
			case <-c.heartbeatStop:
				return
			}
		}
	}()
}

// StopHeartbeat stops automatic heartbeat
func (c *Client) StopHeartbeat() {
	if c.heartbeatStop != nil {
		close(c.heartbeatStop)
		c.heartbeatStop = nil
	}
}

// IsSessionValid checks if session is valid
func (c *Client) IsSessionValid() bool {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.sessionToken == "" || c.sessionExpires == nil {
		return false
	}
	return time.Now().Before(*c.sessionExpires)
}

// GetSessionInfo returns session info
func (c *Client) GetSessionInfo() map[string]interface{} {
	c.mu.Lock()
	defer c.mu.Unlock()

	var expires interface{}
	if c.sessionExpires != nil {
		expires = c.sessionExpires.Format(time.RFC3339)
	}

	return map[string]interface{}{
		"token":    c.sessionToken,
		"expires":  expires,
		"is_valid": c.IsSessionValid(),
	}
}

// Destroy cleans up resources
func (c *Client) Destroy() {
	c.StopHeartbeat()
	c.mu.Lock()
	c.sessionToken = ""
	c.sessionExpires = nil
	c.licenseKey = ""
	c.hwid = ""
	c.mu.Unlock()
}
